src/SharpCompress/Compressor/ADC/ADCBase.cs

```diff
@@ -0,0 +1,210 @@
+        }
+
+        static int GetOffset(byte[] chunk, int position)
+        {
+            switch (GetChunkType(chunk[position]))
+            {
+                case Plain:
+                    return 0;
+                case TwoByte:
+                    return ((chunk[position] & 0x03) << 8) + chunk[position + 1];
+                case ThreeByte:
+                    return (chunk[position + 1] << 8) + chunk[position + 2];
+                default:
+                    return -1;
+            }
+        }
+
+        /// <summary>
+        /// Decompresses a byte buffer that's compressed with ADC
+        /// </summary>
+        /// <param name="input">Compressed buffer</param>
+        /// <param name="output">Buffer to hold decompressed data</param>
+        /// <param name="bufferSize">Max size for decompressed data</param>
+        /// <returns>How many bytes are stored on <paramref name="output"/></returns>
+        public static int Decompress(byte[] input, out byte[] output, int bufferSize = 262144)
+        {
+            return Decompress(new MemoryStream(input), out output, bufferSize);
+        }
+
+        /// <summary>
+        /// Decompresses a stream that's compressed with ADC
+        /// </summary>
+        /// <param name="input">Stream containing compressed data</param>
+        /// <param name="output">Buffer to hold decompressed data</param>
+        /// <param name="bufferSize">Max size for decompressed data</param>
+        /// <returns>How many bytes are stored on <paramref name="output"/></returns>
+        public static int Decompress(Stream input, out byte[] output, int bufferSize = 262144)
+        {
+            output = null;
+
+            if (input == null || input.Length == 0)
+                return 0;
+
+            int start = (int)input.Position;
+            int position = (int)input.Position;
+            int chunkSize;
+            int offset;
+            int chunkType;
+            byte[] buffer = new byte[bufferSize];
+            int outPosition = 0;
+            bool full = false;
+            MemoryStream tempMs;
+
+            while (position < input.Length)
+            {
+                int readByte = input.ReadByte();
+                if (readByte == -1)
+                    break;
+
+                chunkType = GetChunkType((byte)readByte);
+
+                switch (chunkType)
+                {
+                    case Plain:
+                        chunkSize = GetChunkSize((byte)readByte);
+                        if (outPosition + chunkSize > bufferSize)
+                        {
+                            full = true;
+                            break;
+                        }
+                        input.Read(buffer, outPosition, chunkSize);
+                        outPosition += chunkSize;
+                        position += chunkSize + 1;
+                        break;
+                    case TwoByte:
+                        tempMs = new MemoryStream();
```

Comment:
Looks like these  temp MemoryStreams aren't required.  Just write directly to the output or a simple array of two or three bytes?